### PR TITLE
[Quota] Add API to list preset variables

### PR DESCRIPTION
**Pull request by BryanMLima**
### Description

This PR extends the functionality introduced in PR #5909, allowing users to list the preset variables via API, facilitating the creation of the activation rules while creating tariffs for the Quota plugin.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### How Has This Been Tested?

I tested this PR verifying the preset variables for the Quota usage types. Below, are the output for the quota usage type `RUNNING_VMS` with enum 1.

<details><summary>API output</summary>

```
(lab) 🐱 > quota presetvariableslist usagetype=1
{
  "count": 41,
  "variables": [
    {
      "description": "Account owner of the resource.",
      "variable": "account"
    },
    {
      "description": "Role of the account. This field will not exist if the account is a project.",
      "variable": "account.role"
    },
    {
      "description": "Role type of the resource's owner.",
      "variable": "account.role.type"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.role.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.role.name"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.name"
    },
    {
      "description": "Domain owner of the resource.",
      "variable": "domain"
    },
    {
      "description": "Path of the domain owner of the resource.",
      "variable": "domain.path"
    },
    {
      "description": "ID of the resource.",
      "variable": "domain.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "domain.name"
    },
    {
      "description": "Project owner of the resource. This field will not exist if the resource belongs to an account.",
      "variable": "project"
    },
    {
      "description": "ID of the resource.",
      "variable": "project.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "project.name"
    },
    {
      "description": "Type of the record used. Examples for this are: VirtualMachine, DomainRouter, SourceNat, KVM.",
      "variable": "resourceType"
    },
    {
      "description": "Data related to the resource being processed.",
      "variable": "value"
    },
    {
      "description": "ID of the resource.",
      "variable": "value.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "value.name"
    },
    {
      "description": "Host where the VM is running.",
      "variable": "value.host"
    },
    {
      "description": "List of tags of the host where the VM is running (i.e.: [\"a\", \"b\"]).",
      "variable": "value.host.tags"
    },
    {
      "description": "Whether the tag is a rule interpreted in JavaScript.",
      "variable": "value.host.isTagARule"
    },
    {
      "description": "ID of the resource.",
      "variable": "value.host.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "value.host.name"
    },
    {
      "description": "OS of the VM/template.",
      "variable": "value.osName"
    },
    {
      "description": "A list of resources of the account between the start and end date of the usage record being calculated (i.e.: [{zoneId: ..., domainId:...}]).",
      "variable": "value.accountResources"
    },
    {
      "description": "List of tags of the resource in the format key:value (i.e.: {\"a\":\"b\", \"c\":\"d\"}).",
      "variable": "value.tags"
    },
    {
      "description": "Computing offering of the VM.",
      "variable": "value.computeOffering"
    },
    {
      "description": "A boolean informing if the compute offering is customized or not.",
      "variable": "value.computeOffering.customized"
    },
    {
      "description": "ID of the resource.",
      "variable": "value.computeOffering.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "value.computeOffering.name"
    },
    {
      "description": "Template/ISO with which the VM was created.",
      "variable": "value.template"
    },
    {
      "description": "ID of the resource.",
      "variable": "value.template.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "value.template.name"
    },
    {
      "description": "Computing resources consumed by the VM.",
      "variable": "value.computingResources"
    },
    {
      "description": "Current VM's memory (in MiB).",
      "variable": "value.computingResources.memory"
    },
    {
      "description": "Current VM's vCPUs.",
      "variable": "value.computingResources.cpuNumber"
    },
    {
      "description": "Current VM's CPU speed (in MHz).",
      "variable": "value.computingResources.cpuSpeed"
    },
    {
      "description": "The hypervisor where the resource was deployed. Values can be: XenServer, KVM, VMware, Hyperv, BareMetal, Ovm, Ovm3 and LXC.",
      "variable": "value.hypervisorType"
    },
    {
      "description": "Zone where the resource is.",
      "variable": "zone"
    },
    {
      "description": "ID of the resource.",
      "variable": "zone.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "zone.name"
    }
  ]
}
```
</details>

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
clgtm, and lots of test. Just a thought; an integration test seems appropriate for such a profoundly user facing feature, would you say @BryanMLima ?

File: plugins/database/quota/src/main/java/org/apache/cloudstack/api/response/QuotaPresetVariablesItemResponse.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    @SerializedName("description")
+    @Param(description = "description")
+    private String description;
+
+    public QuotaPresetVariablesItemResponse() {
+        super("variables");
+    }
+
+    public void setVariables(String variable) {
```
Comment (sureshanaparti) [inline plugins/database/quota/src/main/java/org/apache/cloudstack/api/response/QuotaPresetVariablesItemResponse.java:36]:
```suggestion
    public void setVariable(String variable) {
```

**Review by sureshanaparti — APPROVED** [review]
changes lgtm

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by BryanMLima** [comment]
> clgtm, and lots of test. Just a thought; an integration test seems appropriate for such a profoundly user facing feature, would you say @BryanMLima ?

@DaanHoogland, got back after the holidays, and looking back, I don't think it needs integration tests. Mainly because there is no interaction with other APIs. The `quotaPresetVariableList` API was created to help users create custom Quota tariffs. Before this API, one would need to keep and maintain a document with the allowed preset variables. With this change, there is an API that is always updated when any changes are made in the code.

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by BryanMLima** [comment]
@blueorangutan package

**Review by GutoVeronezi — APPROVED** [review]
CLGTM

**Review by bernardodemarco — APPROVED** [review]
LGTM, I manually tested this one.

<details>
<summary>Preset variables for quota usage type <code>14</code></summary>

```bash
> quotaPresetVariablesList usagetype=14
{
  "count": 20,
  "variables": [
    {
      "description": "Account owner of the resource.",
      "variable": "account"
    },
    {
      "description": "Role of the account. This field will not exist if the account is a project.",
      "variable": "account.role"
    },
    {
      "description": "Role type of the resource's owner.",
      "variable": "account.role.type"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.role.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.role.name"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.name"
    },
    {
      "description": "Domain owner of the resource.",
      "variable": "domain"
    },
    {
      "description": "Path of the domain owner of the resource.",
      "variable": "domain.path"
    },
    {
      "description": "ID of the resource.",
      "variable": "domain.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "domain.name"
    },
    {
      "description": "Project owner of the resource. This field will not exist if the resource belongs to an account.",
      "variable": "project"
    },
    {
      "description": "ID of the resource.",
      "variable": "project.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "project.name"
    },
    {
      "description": "Type of the record used. Examples for this are: VirtualMachine, DomainRouter, SourceNat, KVM.",
      "variable": "resourceType"
    },
    {
      "description": "Data related to the resource being processed.",
      "variable": "value"
    },
    {
      "description": "A list of resources of the account between the start and end date of the usage record being calculated (i.e.: [{zoneId: ..., domainId:...}]).",
      "variable": "value.accountResources"
    },
    {
      "description": "Zone where the resource is.",
      "variable": "zone"
    },
    {
      "description": "ID of the resource.",
      "variable": "zone.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "zone.name"
    }
  ]
}
```
</details>

<details>
<summary>Preset variables for quota usage type <code>10</code></summary>

```bash
> quotaPresetVariablesList usagetype=10
{
  "count": 20,
  "variables": [
    {
      "description": "Account owner of the resource.",
      "variable": "account"
    },
    {
      "description": "Role of the account. This field will not exist if the account is a project.",
      "variable": "account.role"
    },
    {
      "description": "Role type of the resource's owner.",
      "variable": "account.role.type"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.role.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.role.name"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.name"
    },
    {
      "description": "Domain owner of the resource.",
      "variable": "domain"
    },
    {
      "description": "Path of the domain owner of the resource.",
      "variable": "domain.path"
    },
    {
      "description": "ID of the resource.",
      "variable": "domain.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "domain.name"
    },
    {
      "description": "Project owner of the resource. This field will not exist if the resource belongs to an account.",
      "variable": "project"
    },
    {
      "description": "ID of the resource.",
      "variable": "project.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "project.name"
    },
    {
      "description": "Type of the record used. Examples for this are: VirtualMachine, DomainRouter, SourceNat, KVM.",
      "variable": "resourceType"
    },
    {
      "description": "Data related to the resource being processed.",
      "variable": "value"
    },
    {
      "description": "A list of resources of the account between the start and end date of the usage record being calculated (i.e.: [{zoneId: ..., domainId:...}]).",
      "variable": "value.accountResources"
    },
    {
      "description": "Zone where the resource is.",
      "variable": "zone"
    },
    {
      "description": "ID of the resource.",
      "variable": "zone.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "zone.name"
    }
  ]
}
```
</details>

<details>
<summary>Preset variables for quota usage type <code>6</code></summary>

```bash
> quotaPresetVariablesList usagetype=6
{
  "count": 35,
  "variables": [
    {
      "description": "Account owner of the resource.",
      "variable": "account"
    },
    {
      "description": "Role of the account. This field will not exist if the account is a project.",
      "variable": "account.role"
    },
    {
      "description": "Role type of the resource's owner.",
      "variable": "account.role.type"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.role.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.role.name"
    },
    {
      "description": "ID of the resource.",
      "variable": "account.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "account.name"
    },
    {
      "description": "Domain owner of the resource.",
      "variable": "domain"
    },
    {
      "description": "Path of the domain owner of the resource.",
      "variable": "domain.path"
    },
    {
      "description": "ID of the resource.",
      "variable": "domain.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "domain.name"
    },
    {
      "description": "Project owner of the resource. This field will not exist if the resource belongs to an account.",
      "variable": "project"
    },
    {
      "description": "ID of the resource.",
      "variable": "project.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "project.name"
    },
    {
      "description": "Type of the record used. Examples for this are: VirtualMachine, DomainRouter, SourceNat, KVM.",
      "variable": "resourceType"
    },
    {
      "description": "Data related to the resource being processed.",
      "variable": "value"
    },
    {
      "description": "ID of the resource.",
      "variable": "value.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "value.name"
    },
    {
      "description": "A list of resources of the account between the start and end date of the usage record being calculated (i.e.: [{zoneId: ..., domainId:...}]).",
      "variable": "value.accountResources"
    },
    {
      "description": "List of tags of the resource in the format key:value (i.e.: {\"a\":\"b\", \"c\":\"d\"}).",
      "variable": "value.tags"
    },
    {
      "description": "Size of the resource (in MiB).",
      "variable": "value.size"
    },
    {
      "description": "Provisioning type of the resource. Values can be: thin, sparse or fat.",
      "variable": "value.provisioningType"
    },
    {
      "description": "Disk offering of the volume.",
      "variable": "value.diskOffering"
    },
    {
      "description": "ID of the resource.",
      "variable": "value.diskOffering.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "value.diskOffering.name"
    },
    {
      "description": "Storage where the volume or snapshot is. While handling with snapshots, this value can be from the primary storage if the globalsetting 'snapshot.backup.to.secondary' is false, otherwise it will be from secondary storage.",
      "variable": "value.storage"
    },
    {
      "description": "List of string representing the tags of the storage where the volume is (i.e.: [\"a\", \"b\"]).",
      "variable": "value.storage.tags"
    },
    {
      "description": "Whether the tag is a rule interpreted in JavaScript. Applicable only for primary storages.",
      "variable": "value.storage.isTagARule"
    },
    {
      "description": "Scope of the storage where the volume is. Values can be: ZONE, CLUSTER or HOST. Applicable only for primary storages.",
      "variable": "value.storage.scope"
    },
    {
      "description": "ID of the resource.",
      "variable": "value.storage.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "value.storage.name"
    },
    {
      "description": "The volume format. Values can be: RAW, VHD, VHDX, OVA and QCOW2.",
      "variable": "value.volumeFormat"
    },
    {
      "description": "Zone where the resource is.",
      "variable": "zone"
    },
    {
      "description": "ID of the resource.",
      "variable": "zone.id"
    },
    {
      "description": "Name of the resource.",
      "variable": "zone.name"
    }
  ]
}
```
</details>

File: framework/quota/src/main/java/org/apache/cloudstack/quota/activationrule/presetvariables/Value.java [hunk 2]
```diff
@@ -23,25 +23,75 @@
...
+    @PresetVariableDefinition(description = "Computing offering of the VM.", supportedTypes = {QuotaTypes.RUNNING_VM, QuotaTypes.ALLOCATED_VM})
     private ComputeOffering computeOffering;
+
+    @PresetVariableDefinition(description = "Template/ISO with which the VM was created.", supportedTypes = {QuotaTypes.RUNNING_VM, QuotaTypes.ALLOCATED_VM})
     private GenericPresetVariable template;
+
+    @PresetVariableDefinition(description = "Disk offering of the volume.", supportedTypes = {QuotaTypes.VOLUME})
     private GenericPresetVariable diskOffering;
+
+    @PresetVariableDefinition(description = "Storage where the volume or snapshot is. While handling with snapshots, this value can be from the primary storage if the global" +
```
Comment (bernardodemarco) [inline framework/quota/src/main/java/org/apache/cloudstack/quota/activationrule/presetvariables/Value.java:80]:
```suggestion
    @PresetVariableDefinition(description = "Storage where the volume or snapshot is. While handling with snapshots, this value can be from the primary storage if the global " +
```

Just a small suggestion here.

**Comment by DaanHoogland** [comment]
@BryanMLima will you apply @bernardodemarco 's sugestion https://github.com/apache/cloudstack/pull/8372#discussion_r1616409198 , or can we merge as is?

Also, @BryanMLima this is marked for 4.19.1 but based on main. What to do?

**Comment by BryanMLima** [comment]
@bernardodemarco, thanks for testing it.

> @BryanMLima will you apply @bernardodemarco 's sugestion [#8372 (comment)](https://github.com/apache/cloudstack/pull/8372#discussion_r1616409198) , or can we merge as is?
> 
> Also, @BryanMLima this is marked for 4.19.1 but based on main. What to do?

I changed the milestone to 4.20, we can merge this after the checks have passed.

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by BryanMLima** [comment]
Merging this one based on reviews and manual tests.

**Comment by rohityadavcloud** [comment]
@BryanMLima shouldn't we need to run smoketests on the PR, or do we think they were covered by GHA/simulator smoketests?

**Comment by BryanMLima** [comment]
> @BryanMLima shouldn't we need to run smoketests on the PR, or do we think they were covered by GHA/simulator smoketests?

@rohityadavcloud, the smoke tests were already run in https://github.com/apache/cloudstack/pull/8372#issuecomment-1936849088, and the only test error was not related to the changes in this PR (further commits did not add any major changes). Furthermore, this patch only introduces new features to the Quota plugin, it does not change any behaviour that would break the smoke tests.

**Pull request merged**